sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,301 @@
+          } else {
+            Array(0)
+          }
+
+          for (i <- 0 until leftMapIdStartIndices.length;
+               j <- 0 until rightMapIdStartIndices.length) {
+            val leftEndMapId = if (i == leftMapIdStartIndices.length - 1) {
+              getMappersNum(left)
+            } else {
+              leftMapIdStartIndices(i + 1)
+            }
+            val rightEndMapId = if (j == rightMapIdStartIndices.length - 1) {
+              getMappersNum(right)
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // TODO we may can optimize the sort merge join to broad cast join after
+            // we get the raw data size of per partition,
+            val leftSkewedReader =
+              SkewedShufflePartitionReader(
+                left, partitionId, leftMapIdStartIndices(i), leftEndMapId)
+
+            val rightSkewedReader =
+              SkewedShufflePartitionReader(right, partitionId,
+                rightMapIdStartIndices(j), rightEndMapId)
+            subJoins += SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+              leftSkewedReader, rightSkewedReader)
+          }
+        }
+      }
+      logDebug(s"number of skewed partitions is ${skewedPartitions.size}")
+      if (skewedPartitions.size > 0) {
+        val partitionIndices = estimatePartitionStartEndIndices(
+          getStatistics(left), skewedPartitions)
+        val optimizedSmj = smj.transformDown {
+          case sort: SortExec if (sort.child.isInstanceOf[QueryStageExec] &&
+            ShuffleQueryStageExec.isShuffleQueryStageExec(sort.child)) => {
+            val partialReader = PartialShuffleReader(
+                sort.child.asInstanceOf[QueryStageExec], partitionIndices)
+            sort.copy(child = partialReader)
+          }
+        }
+        subJoins += optimizedSmj
+        UnionExec(subJoins)
+      } else {
+        smj
+      }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.getConf(SQLConf.ADAPTIVE_EXECUTION_SKEWED_JOIN_ENABLED)) {
+      return plan
+    }
+
+    def collectShuffleStages(plan: SparkPlan): Seq[ShuffleQueryStageExec] = plan match {
+      case _: LocalShuffleReaderExec => Nil
+      case _: SkewedShufflePartitionReader => Nil
+      case _: PartialShuffleReader => Nil
+      case _: CoalescedShuffleReaderExec => Nil
+      case stage: ShuffleQueryStageExec => Seq(stage)
+      case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) => Seq(stage)
+      case _ => plan.children.flatMap(collectShuffleStages)
+    }
+
+    val shuffleStages = collectShuffleStages(plan)
+
+    if (shuffleStages.length == 2) {
+      // Currently we only support handling skewed join for 2 table join.
+      handleSkewJoin(plan)
+    } else {
+      plan
+
+    }
+  }
+}
+
+case class PartialShuffleReader(
+    child: QueryStageExec, partitionRanges: Array[(Int, Int)]) extends UnaryExecNode {
```

Comment:
I think it's better to pass a `excludedPartitions: Set[Int]` here, and then calculate the partition index ranges inside this class.

Reply:
If so, we may re-calculated the `mapstatistics` of `QueryStage` in this class?